refactor!: restructure project for better DX and UX

sseemayer opened the pull request:
This PR is a major change to the internal structure and API interface in preparation for a 1.0 release of the crate.

* feat: add attachment support
* feat: add custom icon support
* feat: add history tracking editing

BREAKING CHANGE: Different database types and how they are accessed. Removes the `Node`, `NodeRef`, etc concepts.
BREAKING CHANGE: Completely replaced error types to make them more granular
BREAKING CHANGE: Structure of output generated by serde::Serialize has changed

sseemayer committed:
refactor: move xml_db into format, where it is used

sseemayer committed:
refactor: improve hmac block stream error handling

sseemayer committed:
chore: export ChallengeResponseKey

sseemayer committed:
feat: kdb v1 parsing

sseemayer committed:
fix: fixup warnings

sseemayer committed:
feat: begin rewrite of xml_db

sseemayer committed:
fix: root initialization

sseemayer committed:
refactor: split xml_db into modules

sseemayer committed:
refactor: move custom (de)serializers into module

sseemayer committed:
test: memory protection

sseemayer committed:
feat: add CustomData serialization/deserialization

sseemayer committed:
fix: make color serialize into upper case hex to match keepass xml

sseemayer committed:
test: meta serialization/deserialization tests

sseemayer committed:
refactor: wip

sseemayer committed:
feat: wip kdbx3

sseemayer committed:
feat: add xml_db types for entry and group

droidmonkey commented:
FWIW, I had copilot agent analyze and propose a solution to transactional edits to the database. It also proposed a JSON & API based method for these transactions. This was a thought project on how we could implement immutable database/group/entry sharing in KeePassXC and move all editing into a centralized place. https://github.com/droidmonkey/keepassxc/pull/6

sseemayer committed:
feat: kdbx3 re-enabled, pending testing

sseemayer committed:
feat: add additional fields

sseemayer committed:
feat: re-enable kdbx4

sseemayer committed:
test: add test to ensure we allow empty XML fields with Option<T>

sseemayer committed:
fix: allow some empty tags to deserialize as None

sseemayer committed:
feat: more optional fields

sseemayer committed:
chore: annotate TOML entries as such, not as ignored Rust tests

sseemayer committed:
feat: implement XmlBridge for Meta

sseemayer committed:
feat: connect meta into database XmlBridge

sseemayer committed:
docs: xml_db design

sseemayer committed:
fix: change enable_auto_type and enable_searching type signature using info from schema

sseemayer committed:
chore: remove test.xml

sseemayer committed:
fix: adjust more types to fit the XML schema

Especially, the Times struct can never have other values than the
hardcoded ones.

sseemayer committed:
feat: convert xml_db group and entry hierarchy into db

sseemayer committed:
docs: notes about what kind of attachments are used where

sseemayer committed:
chore: remove olddb

sseemayer committed:
refactor: redesign attachment system

Have just one user-facing Attachment type, with the parsers handling the
translation of different ways to specify attachments into this type.

sseemayer committed:
refactor: finish renaming header_attachments -> attachments

sseemayer committed:
chore: mark some types as #[non_exhaustive]

sseemayer committed:
feat: handle attachment refs and autotype bridging

sseemayer committed:
feat: convert xml_db history for entries

sseemayer committed:
feat: mutable iteration over all entries and groups

sseemayer committed:
feat: num_entries, num_groups

sseemayer committed:
chore: fix some warnings

sseemayer committed:
docs: guidance on how to use new API

sseemayer committed:
chore: document and tweak API

sseemayer committed:
feat: convenience methods for entry fields

sseemayer committed:
docs: also use convenience methods in theh save example

sseemayer committed:
feat: kdbx4 writing

sseemayer committed:
fix: customdata conversion errors on default features

sseemayer committed:
test: re-enable file_read_tests

sseemayer committed:
test: re-enable entry_tests

sseemayer committed:
chore: fix some unused warnings due to features

sseemayer committed:
chore: remove AGENTS.md

sseemayer committed:
test: re-enable large file roundtrip tests and kdbx4 tests

sseemayer committed:
feat: access parent group in entry and group, moving to new parent

sseemayer committed:
feat: merge implementation (pending tests)

sseemayer committed:
feat: convenience methods for testing and editing

sseemayer committed:
chore: begin translating merge tests

sseemayer committed:
fix: deleted_objects cannot distinguish between entry and group

sseemayer committed:
feat: add GroupTrack, history-tracking removal

sseemayer committed:
chore: wip merge tests

sseemayer committed:
chore: merge tests now compile

sseemayer committed:
chore!: remove kp-dump-json

with the new internal structure, creating a JSON dump of the database is
less straightforward and it's better to rethink how to reimplement such
a feature than getting a half-baked solution

BREAKING: cli command kp-dump-json removed

sseemayer committed:
fix: update kp-purge-history to new API

sseemayer committed:
feat: re-add TOTP support, fix kp-show-otp

sseemayer committed:
chore: documentation for totp

sseemayer committed:
fix: some of the merge tests

sseemayer committed:
refactor: re-write merge (again)

sseemayer committed:
chore: fix merge tests

sseemayer committed:
chore: move History::is_ordered into merge tests

sseemayer committed:
fix: wasm test

sseemayer committed:
fix: wasm time

sseemayer committed:
docs: more elaborate open_database and save_database examples

showcase various ways of using the new API for interfacing with the
database structure

sseemayer committed:
chore: remove examples from README

sseemayer committed:
chore: apply clippy hints

Its-Just-Nans committed:
feat: improve create security on unchecked array access (#285)

make open_kdbx4_fuzzing complete more quickly so it can always be run by
using a database with very few AES rounds for key derivation

---------

Co-authored-by: Stefan Seemayer <stefan@seemayer.de>

sseemayer committed:
chore: get rid of sleeps in merge tests (not needed)

sseemayer committed:
fix: go back to second-accuracy time

sseemayer committed:
fix: do not create XML fields if they are none

sseemayer committed:
chore: additional tests

sseemayer committed:
chore: additional tests

sseemayer committed:
feat: entry CustomData serialization and deserialization

sseemayer committed:
chore: also test memory protection serialization/deserialization

sseemayer committed:
fix: attachment serialization/deserialization

sseemayer committed:
feat: re-add serde serialization support

sseemayer committed:
chore: small tweaks

sseemayer committed:
feat: custom icon serialization/deserialization

sseemayer committed:
docs: increase doc coverage

sseemayer committed:
chore: more tests

sseemayer committed:
chore: fix formatting

sseemayer committed:
chore: more idiomatic kp-purge-history

sseemayer committed:
chore: fix clippy warnings with -W clippy::indexing_slicing

sseemayer committed:
chore: configure clippy to warn on indexing_slicing

sseemayer committed:
chore: clippy lints

sseemayer committed:
chore: docs for examples

sseemayer committed:
chore: module-level docs for test files

sseemayer committed:
chore: fix clippy lints

sseemayer commented:
@Its-Just-Nans , thanks for the suggestion! 

I have added these lints to `Cargo.toml`: 

```toml
[lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = ['cfg(tarpaulin_include)'] }
missing_docs = "warn"
rust_2018_idioms = "warn"

[lints.clippy]
indexing_slicing = "deny"
missing_panics_doc = "deny"
unwrap_used = "deny"
expect_used = "deny"
cast_lossless = "deny"
cargo = { level = "warn", priority = -1 }
```

and fixed many many lint errors. There are many cases where `expect` and `unwrap` are used in a way that cannot cause a panic, and they are all justified in comments next to `#[allow(clippy::unwrap_used)]` now.

yoyosource commented:
After trying out the current released version I really like this library. Though it bugged me that I was not able to connect the attachments inside of the Header to a specific Entry. Which is why I tried out the current rev of this branch and after some refactoring I was no longer able to read my KDBX4 file. The `Database::open` complained about the `IsExpanded` property not being present. It would be great to have a sensible default (most likely `false`) for this XML Field to enable parsing of KDBX4 files that dont have this Field present.

sseemayer committed:
fix: make IsExpanded optional

yoyosource commented:
Here you go:
[IsExpandedTest.kdbx.zip](https://github.com/user-attachments/files/25292662/IsExpandedTest.kdbx.zip)
The password is `TestTestTest` it only contains 1 Group and nothing else. I created this KDBX4 with Strongbox and I tried to open it with the latest commit of this branch and still got this error:
```
Custom { kind: Other, error: Parse(KDB4(Xml(Xml(Custom("missing field `IsExpanded`"))))) }
```

sseemayer committed:
fix: is_expanded default + test

sseemayer committed:
fix(xml): allow Group and Entry tags to appear in any order

yoyosource commented:
@sseemayer Thanks for the fix. I tried the new commit locally and it works as expected now. When I find another issue I will write here.

One more things how can I convert an Entry from a history to a EntryRef so I can access the stored attachments? Or is that impossible by design/spec?

sseemayer committed:
chore: re-add test for tab in keyfile

sseemayer committed:
feat: restoring historical attachments

sseemayer committed:
chore: add clippy allow

sseemayer committed:
fix: also load historical attachment name

sseemayer commented:
> One more things how can I convert an Entry from a history to a EntryRef so I can access the stored attachments? Or is that impossible by design/spec?

@yoyosource  there is a bit of weirdness necessary because `EntryRef` must point to a "live" Entry in the database, and the History and all historical entries is contained within a live Entry. I have now added a `restore_entry` function that allows you to promote a historical version of an Entry to a full Entry, and get whatever data from there.

See [this test](https://github.com/sseemayer/keepass-rs/blob/16361b4f59447ec113cc0a6bf610cc5400ad3566/tests/entry_tests.rs#L127) for an example on how to use.

yoyosource commented:
@sseemayer Wouldnt it be possible to allow accessing the Attachments on an Entry instead of only on an EntryRef instead? Since the data needs to be there regardless of having an EntryRef or an Entry. And then I would access the Attachment from the Database manually.

sseemayer committed:
feat: attachment_ids , entry_ids, group_ids

yoyosource commented:
Looks good thanks. I will try it out in a bit but the test looks like what I need.

yoyosource commented:
No actually there is some other issue here. Somehow it sometimes loads the Attachments from the previous Entry instead of the current one. I am still trying to find out why this happens.

sseemayer committed:
fix: mismatched attachments

thanks to @yoyosource for reporting the issue and a fix

sseemayer commented:
@yoyosource thanks a lot for the report and fix!

sseemayer commented:
after discussing with @louib , this PR will not be merged as-is, but rather re-implemented as a series of more reviewable changes, the first one (#290) now in draft stage. The end result will likely be very similar to this branch, though.

sseemayer commented:
This has now been completely replaced by #291 #292 #293 #294
